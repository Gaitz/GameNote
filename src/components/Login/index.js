import React, { useContext, useEffect } from "react"
import { useCreateStore, storeContext } from "@hooks"

export default function Login() {
  return <Parent></Parent>
}

const add = {
  type: "ADD",
  value: 2,
}

const actionTypes = {
  ADD: "ADD",
}

const reducerInitialState = {
  sum: 0,
}

const sumReducer = (old_state = reducerInitialState, action) => {
  const { type } = action
  switch (type) {
    case actionTypes.ADD:
      return { ...old_state, sum: old_state.sum + action.value }
    default:
      return old_state
  }
}

const initialState = {
  initialString: "",
  initialNumber: 123,
  initialObject: { nested: "nestedObject" },
}

function Parent() {
  const Store = useCreateStore(sumReducer, initialState)
  return (
    <Store>
      <div>hello</div>
      <Child></Child>
    </Store>
  )
}

let callTimes = 0

function Child() {
  const store = useContext(storeContext)

  useEffect(() => {
    console.log(callTimes)
    callTimes++
  })

  useEffect(() => {
    new Promise((resolve) => {
      setTimeout(() => {
        store.dispatch(add)
        return resolve()
      }, 0)
    })
  }, [])

  return <p>{console.log(store.state)}</p>
}
