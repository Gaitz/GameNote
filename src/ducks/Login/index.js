import React from "react"
import { storeContext } from "@game-note/ducks/App"

export default function Login() {
  const store = React.useContext(storeContext)
  return (
    <>
      <h2>Login Component</h2>
      <p>{store.state.sayHi}</p>
      <p>{JSON.stringify(store.state)}</p>
    </>
  )
}
