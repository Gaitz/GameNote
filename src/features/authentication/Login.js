import React from "react"
import styles from "./Login.style.css"
import firebaseService from "game-note/services/firebaseService"

export default function Login() {
  const [formState, setFormState] = React.useState({
    isPasswordVisible: false,
    errorMessage: null,
  })

  const [loginData, setLoginData] = React.useState({
    email: null,
    password: null,
  })

  return (
    <article className={styles.loginComponent}>
      <form
        onSubmit={(event) => {
          event.preventDefault()
          if (formState.errorMessage) {
            event.target.reset()
            return
          }
          if (loginData.email && loginData.password) {
            firebaseService.auth
              .signInWithEmailAndPassword(loginData.email, loginData.password)
              .catch((error) => {
                console.error(error)
                setFormState({ ...formState, errorMessage: error.message })
              })
          }
        }}
        onReset={() => {
          setFormState({ ...formState, errorMessage: null })
        }}
      >
        <div className={styles.inputBlock}>
          <label htmlFor="email">Email:</label>
          <input
            autoFocus
            id="email"
            name="email"
            type="email"
            required
            autoComplete="username"
            onChange={(event) =>
              setLoginData({ ...loginData, email: event.target.value })
            }
          ></input>
        </div>
        <div className={styles.inputBlock}>
          <div className={styles.passwordLabelBlock}>
            <label htmlFor="password">Password:</label>
            <span
              className={styles.toggleVisibleButton}
              onClick={(event) => {
                event.preventDefault()
                setFormState({
                  ...formState,
                  isPasswordVisible: !formState.isPasswordVisible,
                })
              }}
            >
              toggle visible
            </span>
          </div>
          <input
            id="password"
            name="password"
            type={formState.isPasswordVisible ? "text" : "password"}
            required
            autoComplete="current-password"
            onChange={(event) =>
              setLoginData({ ...loginData, password: event.target.value })
            }
          ></input>
        </div>
        <button type="submit">Login</button>
        {formState.errorMessage ? (
          <button type="reset" className={styles.errorMessageBlock}>
            {formState.errorMessage}
          </button>
        ) : null}
      </form>
    </article>
  )
}
