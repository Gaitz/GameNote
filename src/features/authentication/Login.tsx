import { useMemo, useState } from "react"
import { Button, Fade, Flex, Heading, SimpleGrid, Text } from "@chakra-ui/react"
import { useTranslation } from "next-i18next"
import {
  useFirebaseAuthLogin,
  EmailLogin
} from "game-note/features/authentication"
import { Loading } from "game-note/shared/components"
import { ChangeLanguage } from "game-note/features/internationalization"

export function Login() {
  const { t } = useTranslation("login")

  const { isLoading, handleGitHubSignIn } = useFirebaseAuthLogin()

  const [emailInputVisible, toggleEmailInputVisible] = useState(false)

  const loginInputMarginRight = useMemo(
    () => ({
      base: "0",
      md: "calc(30vw - 117.5px)",
      lg: "calc(34vw - 117.5px)"
    }),
    []
  )

  if (isLoading) {
    return <Loading />
  }

  return (
    <>
      <Flex mr={loginInputMarginRight} justifyContent="flex-end">
        <ChangeLanguage />
      </Flex>
      <Flex
        mt={["8", "32"]}
        flexDirection="row"
        alignItems="flex-start"
        justifyContent={{
          base: "center",
          md: "flex-end"
        }}
        flexWrap="wrap"
      >
        <Heading
          as="h1"
          flexBasis={{ base: "100%", md: "45%" }}
          fontSize={["3xl", "5xl"]}
          display="flex"
          flexDirection="column"
          alignItems="center"
        >
          <Text variant="heading" align="center">
            {t("welcome")}
          </Text>
          Game Note
        </Heading>
        <SimpleGrid
          spacing="5"
          mt={{ base: "5", md: "0" }}
          mr={loginInputMarginRight}
        >
          <Button onClick={handleGitHubSignIn}>GitHub {t("login")}</Button>
          <Button onClick={() => toggleEmailInputVisible(!emailInputVisible)}>
            Email {t("login")}
          </Button>
          <Fade in={emailInputVisible}>
            <EmailLogin />
          </Fade>
        </SimpleGrid>
      </Flex>
    </>
  )
}
