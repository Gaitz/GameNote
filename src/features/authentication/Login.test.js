import React from "react"
import { Provider } from "react-redux"
import { fireEvent, render } from "@testing-library/react"
import "@testing-library/jest-dom"
import configureStore from "redux-mock-store"
import Login from "./Login"

const mockStore = configureStore()

test(
  "test password toggle visible",
  () => {
    const initialState = {
      authentication: {
        user: null,
        error: null,
        isPending: false
      }
    }
    const store = mockStore(initialState)
    const { container, getByTestId, getByLabelText } = render(<Provider store={store}>
      <Login />
    </Provider>)
    expect(container).toMatchSnapshot()

    const toggleVisibleButton = getByTestId("toggleVisible")
    expect(toggleVisibleButton).toBeTruthy()
    fireEvent.click(toggleVisibleButton)

    expect(getByLabelText("Password:")).toHaveAttribute(
      "type",
      "text"
    )
  }
)
