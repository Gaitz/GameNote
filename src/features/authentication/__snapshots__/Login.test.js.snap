// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test password toggle visible 1`] = `
<div>
  <h1>
    Game Note
  </h1>
  <article
    class="loginComponent"
  >
    <form>
      <div
        class="inputBlock"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          autocomplete="username"
          id="email"
          name="email"
          required=""
          type="email"
        />
      </div>
      <div
        class="inputBlock"
      >
        <label
          for="password"
        >
          Password:
        </label>
        <div
          class="passwordInputBlock"
        >
          <input
            autocomplete="current-password"
            id="password"
            name="password"
            required=""
            type="password"
          />
          <img
            alt="Click to toggle password visible"
            class="toggleVisibleButton"
            data-testid="toggleVisible"
            src="[object Object]"
          />
        </div>
      </div>
      <button
        class="loginButton"
        type="submit"
      >
        Login
      </button>
    </form>
  </article>
</div>
`;
